### ISSUE TITLE
ConvertTo-Xml doesn't limit the depth for (recursive) HashTables (Crash)

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#7290]
Steps to reproduce
------------------
Insert the following commands

```powershell
$a = @{Key = "Test"}
$a.Parent = $a
$a | ConvertTo-Xml -As String
```
Expected behavior
-----------------
I would expect an XML string with the output limited to the default depth of **3**:

```powershell
<Property Name="Key" Type="System.String">Key</Property>
<Property Name="Value" Type="System.String">Test</Property>
<Property Name="Key" Type="System.String">Parent</Property>
<Property Name="Value" Type="System.Collections.Hashtable">
  <Property Name="Key" Type="System.String">Key</Property>
  <Property Name="Value" Type="System.String">Test</Property>
  <Property Name="Key" Type="System.String">Parent</Property>
  <Property Name="Value" Type="System.Collections.Hashtable">
    <Property Name="Key" Type="System.String">Key</Property>
    <Property Name="Value" Type="System.String">Test</Property>
    <Property Name="Key" Type="System.String">Parent</Property>
    <Property Name="Value" Type="System.String">end</Property>
  </Property>
</Property>
```

Actual behavior
---------------
Instead, it first shows:

> Process is terminating due to StackOverflowException.

And a few seconds later the whole shell is terminated.

- Note: the same issue appears when the depth is explicitly given (e.g. `-Depth 3`)

**PowerShell Windows**
A similar issue exist in PowerShell Windows (5.1) which reports:

> ConvertTo-Xml : Stream was too long.
> At line:1 char:6
> + $a | ConvertTo-Xml
> +      ~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (:) [ConvertTo-Xml], IOException
>     + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Commands.ConvertToXmlCommand

But does *not* crash.

Environment data
----------------

```powershell
Name                           Value
----                           -----
PSVersion                      6.0.2
PSEdition                      Core
GitCommitId                    v6.0.2
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

ConvertTo-Xml version: 3.1.0.0
```

**Comment by iRon7** [comment]
I wonder whether I should split this issue in:

1. `ConvertTo-Xml` *cmdlet*: doesn't limit the depth for (recursive) HashTables
2. PowerShell Process terminates when the stream is too long

**Comment by iSazonov** [comment]
@iRon7 Thanks for your report!

**Comment by IISResetMe** [comment]
@iRon7 it's not just `[hashtable]` - `ConvertTo-Xml` does this (ignores depth and writes members of) a number of built-in collection types, inclusing any `Dictionary`, `Stack` or `List` type.

**Comment by iSazonov** [comment]
The fix is seems a breaking change. 
/cc @SteveL-MSFT

**Comment by SteveL-MSFT** [comment]
@iSazonov to be clear, the breaking change here is that `-Depth` would now be enforced when previously it did not?

**Comment by iRon7** [comment]
Yes, that is how I see it.

Op do 20 dec. 2018 20:07 schreef Steve Lee <notifications@github.com:

**Comment by iSazonov** [comment]
@SteveL-MSFT Yes, again Depth :-) Currently it is ignored for nested properties and containers. I think it must fit *-Json cmdlets.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this.  Concern is introducing a breaking change for existing scripts that have deep nested hashtables that work today and would be different if we started enforcing `-Depth`.  Recommendation is to keep current default behavior with hashtable except test if sufficient stack is available and throw an exception.  If `-Depth` is explicitly specified, we should change the cmdlet to enforce that depth for hashtables.  This should alleviate breaking change concerns.

**Comment by danekan** [comment]
Regarding the last comment and breaking change status, it's incorrect to assess this risk as only impacting hashtables, it doesn't work for many other examples.

**Comment by IISResetMe** [comment]
@danekan correct, if you look at the [initial part of the solution](https://github.com/PowerShell/PowerShell/pull/8476/files#diff-f5f7d9a73edf98d4af93244f28511cde) you'll see that it's not hashtable/dictionary-specific

**Comment by iRon7** [comment]
The issue still exists in PowerShell `7.4.2`

**Comment by iRon7** [comment]
The issue still exist in:

```PowerShell
Name                           Value
----                           -----
PSVersion                      7.4.6
PSEdition                      Core
GitCommitId                    7.4.6
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by iRon7** [comment]
The issue still exist but improved in a way that it now (version `7.5.1`) reports (after a long time):

> ConvertTo-Xml: Stream was too long.

There is still no way to control the depth or get some results to a certain depth
